File: ios/Runner/WebViewAssetBaseUrl.swift
import Flutter

/// Exposes the on-disk `flutter_assets/` URL for Dart to pass as `loadData`'s `baseUrl`.
class WebViewAssetBaseUrl {

    private static let channelName = "com.linagora.tmail/webview_asset_base_url"
    private static let flutterAssetsBaseUrlMethod = "flutterAssetsBaseUrl"

    func register(_ binaryMessenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(
            name: Self.channelName,
            binaryMessenger: binaryMessenger
        )
        channel.setMethodCallHandler { [weak self] call, result in
            switch call.method {
            case Self.flutterAssetsBaseUrlMethod:
                result(self?.flutterAssetsBaseUrl())
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    private func flutterAssetsBaseUrl() -> String? {
        // lookupKey(forAsset:) joins "<flutterAssetsDir>/<asset>"; trimming
        // the known asset name back off recovers the directory name, since
        // the underlying flutterAssetsName isn't exposed directly.
        let probeKey = FlutterDartProject.lookupKey(forAsset: "AssetManifest.json")
        guard let flutterAssetsDirName = probeKey
            .range(of: "/AssetManifest.json")
            .map({ String(probeKey[probeKey.startIndex..<$0.lowerBound]) }) else {
            return nil
        }
        let flutterAssetsUrl = Bundle.main.bundleURL.appendingPathComponent(
            flutterAssetsDirName,
            isDirectory: true
        )
        return flutterAssetsUrl.absoluteString
    }
}

File: android/app/src/main/kotlin/com/linagora/android/tmail/WebViewAssetBaseUrl.kt
package com.linagora.android.tmail

import io.flutter.FlutterInjector
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel

/** Exposes the on-disk `flutter_assets/` URL for Dart to pass as `loadData`'s `baseUrl`. */
class WebViewAssetBaseUrl {

    fun register(flutterEngine: FlutterEngine) {
        MethodChannel(
            flutterEngine.dartExecutor.binaryMessenger,
            CHANNEL,
        ).setMethodCallHandler { call, result ->
            when (call.method) {
                FLUTTER_ASSETS_BASE_URL_METHOD -> result.success(flutterAssetsBaseUrl())
                else -> result.notImplemented()
            }
        }
    }

    private fun flutterAssetsBaseUrl(): String {
        val flutterAssetsDir = FlutterInjector.instance().flutterLoader().findAppBundlePath()
        return "$ANDROID_ASSET_URL_PREFIX$flutterAssetsDir/"
    }

    private companion object {
        const val CHANNEL = "com.linagora.tmail/webview_asset_base_url"
        const val FLUTTER_ASSETS_BASE_URL_METHOD = "flutterAssetsBaseUrl"
        const val ANDROID_ASSET_URL_PREFIX = "file:///android_asset/"
    }
}
